### ISSUE TITLE
CORS middleware not working anymore

### ISSUE THREAD

**Issue by schnitzelbub** [bocadilloproject/bocadillo#188]
Precondition: `CORS` middleware is enabled as described in [quickstart](../blob/305d13ba87f60195a8433afb0d6016af9301c36f/docs/getting-started/snippets/app.py#L130:L136).

**Expected behavior**
Requests sent with HTTP header `Origin` should be replied with according `Access-Control-*` headers.

**Actual behavior**
No CORS related headers are added anymore.

**To Reproduce**
Send a request like this:

```http
GET /courses/ HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: localhost:8000
Origin: localhost:8000
User-Agent: HTTPie/1.0.2
```

and the response will be like that:

```http
HTTP/1.1 200 OK
content-length: 2
content-type: application/json
date: Tue, 05 Mar 2019 18:28:08 GMT
server: uvicorn

[]
```

**Specs**
- OS: Linux 4.20 x86_64
- Python Version: 3.7.2
- Bocadillo Version: 0.12.4

**Additional context**
This used to work in a previous version (0.11.x I believe).

**Comment by florimondmanca** [comment]
Hi @schnitzelbub,

I investigated this and the quickstart example is actually misleading. As described in the [CORS middleware documentation](https://bocadilloproject.github.io/guides/agnostic/asgi-middleware.html#cors), CORS has restrictive defaults, and one has to explicitly pass a non-empty `allow_origins` to `cors_config`.

Can you confirm the following example app behaves as expected?

```python
from bocadillo import App

app = App(enable_cors=True, cors_config={"allow_origins": ["*"]})

@app.route("/")
async def index(req, res):
    res.text = "Hello"

if __name__ == "__main__":
    app.run(debug=True)
```

Example request:

```python
>>> import requests
>>> r = requests.get("http://localhost:8000", headers={"Origin": "example.com"})
>>> r.headers
{'server': 'uvicorn', 'date': 'Tue, 05 Mar 2019 21:36:01 GMT', 'content-type': 'text/plain', 'content-length': '5', 'access-control-allow-origin': '*'}
```

If so, I will update the documentation to include the `allow_origins` parameter. :+1:

**Comment by schnitzelbub** [comment]
Hi @florimondmanca 

thanks for your quick reply. I can confirm that your example works, but after adding a `Recipe` to the application it fails.
Here's a complete repro using your example including a test:

```python3
import pytest
from bocadillo import App, Recipe

app = App(
    enable_cors=True,
    cors_config={"allow_origins": ["example.com", "localhost:8001", "ietf.org"]},
)


@app.route("/")
async def index(req, res):
    res.text = "Hello"


courses = Recipe("courses")


@courses.route("/")
class CoursesList:
    async def get(self, req, res):
        res.media = ["1", "2"]


app.recipe(courses)

if __name__ == "__main__":
    app.run(debug=True)


@pytest.mark.parametrize(
    ["path", "origin", "expected", "expected_body"],
    [
        ("/", "localhost:8001", "localhost:8001", "Hello"),
        ("/", "ietf.org", "ietf.org", "Hello"),
        ("/", "unknown.org", None, "Hello"),
        ("/courses/", "localhost:8001", "localhost:8001", '["1", "2"]'),
        ("/courses/", "ietf.org", "ietf.org", '["1", "2"]'),
        ("/courses/", "unknown.org", None, '["1", "2"]'),
    ],
)
def test_cors_headers(path: str, origin: str, expected: str, expected_body: str):
    res = app.client.request("GET", path, headers={"origin": origin})

    if not expected:  # unknown origin -> no allow-origin header
        assert "access-control-allow-origin" not in res.headers
    else:  # "allowed origin -> allow-origin header"
        assert "access-control-allow-origin" in res.headers
        assert res.headers.get("access-control-allow-origin") == expected

    assert res.text == expected_body, "assert response body"
```
Tests 4 and 5 for the allowed origins are failing on my end.

**Comment by florimondmanca** [comment]
Okay, I see, thanks for the thorough example and tests.

From 0.12, recipes are treated like mounted (separate) apps, which means that middleware applied to the root App is only mounted there and not on recipes.

There are open issues to deal with this (#117, #168), and we may also need a way to set default configuration for all apps (including recipes).

In the meantime, you should be able to make it work by also setting CORS configuration on the recipe. Can you try that out?

**Comment by schnitzelbub** [comment]
@florimondmanca thanks, that would do it for now:

```python3
import pytest
from bocadillo import App, Recipe

cors_config = {"allow_origins": ["example.com", "localhost:8001", "ietf.org"]}

app = App(enable_cors=True, cors_config=cors_config)


@app.route("/")
async def index(req, res):
    res.text = "Hello"


courses = Recipe(
    name="courses",
    # workaround for https://github.com/bocadilloproject/bocadillo/issues/188
    enable_cors=True,
    cors_config=cors_config,
)


@courses.route("/")
class CoursesList:
    async def get(self, req, res):
        res.media = ["1", "2"]


app.recipe(courses)

if __name__ == "__main__":
    app.run(debug=True)


@pytest.mark.parametrize(
    ["path", "origin", "expected", "expected_body"],
    [
        ("/", "localhost:8001", "localhost:8001", "Hello"),
        ("/", "ietf.org", "ietf.org", "Hello"),
        ("/", "unknown.org", None, "Hello"),
        ("/courses/", "localhost:8001", "localhost:8001", '["1", "2"]'),
        ("/courses/", "ietf.org", "ietf.org", '["1", "2"]'),
        ("/courses/", "unknown.org", None, '["1", "2"]'),
    ],
)
def test_cors_headers(path: str, origin: str, expected: str, expected_body: str):
    res = app.client.request("GET", path, headers={"origin": origin})

    if not expected:  # unknown origin -> no allow-origin header
        assert "access-control-allow-origin" not in res.headers
    else:  # allowed origin -> allow-origin header"
        assert "access-control-allow-origin" in res.headers
        assert res.headers.get("access-control-allow-origin") == expected

    assert res.text == expected_body, "assert response body"
```

But as CORS only applies to a combination of `protocol prefix`, `hostname` and `port` but not to paths there should - at least for this setting - be only one configuration for the whole application including all its subcomponents. Overwriting it in a subcomponent (`recipe`) wouldn't make sense, would it?

**Comment by florimondmanca** [comment]
@schnitzelbub I've proposed a fix in #195 which passes the tests you listed in your previous answer. Feel free to take a look. :-)

**Comment by schnitzelbub** [comment]
@florimondmanca thanks, this looks solid. I just renamed some of the variables in the test to be more in sync with your documentation changes and thus provided a PR #197 on top of your PR ;-)
